src/scanner.rs

```diff
@@ -1,63 +1,100 @@
-use global_state::GlobalState;
+use rox::Rox;
 use token::Token;
+use token_type::TokenType;
 
-use std::fs::File;
-use std::io;
-use std::io::{BufRead, Write};
-use std::io::prelude::*;
-use std::process;
-
+#[derive(Debug)]
 pub struct Scanner {
     source: String,
+    tokens: Vec<Token>,
+    start: usize,
+    current: usize,
+    line: usize,
 }
 
 impl Scanner {
-    fn scan_tokens(&self) -> Vec<Token> {
-        self.source
-            .split(" ")
-            .map(|n| Token::new(n.to_string())).collect::<Vec<Token>>()
-    }
-    pub fn run_file(file_name: &String, state: &mut GlobalState) {
-        println!("source: {}", file_name);
-        let mut f = File::open(file_name).expect("file not found");
-        let mut contents = String::new();
-        f.read_to_string(&mut contents)
-            .expect("could not parse file");
-        Scanner::run(&contents, state);
-        if state.had_error {
-            process::exit(65);
+    pub fn new(source: String) -> Scanner {
+        Scanner {
+            source: source,
+            tokens: Vec::new(),
+            start: 0,
+            current: 0,
+            line: 1,
+        }
+    }
+    pub fn scan_tokens(&mut self, rox: &mut Rox) {
+        while !self.is_at_end() {
+            self.start = self.current;
+            self.scan_token(rox); // uses Rox::error
         }
+        self.add_token(TokenType::Eof);
     }
-    pub fn run_prompt(state: &mut GlobalState) {
-        let mut line = String::new();
-        let stdin = io::stdin();
-        loop {
-            print!("> ");
-            io::stdout().flush().unwrap();
-            stdin.lock().read_line(&mut line).expect("syntax error");
-            if line.trim() == "exit" {
-                break;
+    fn is_at_end(&self) -> bool {
+        self.current >= self.get_source_chars().len()
+    }
+    pub fn print_tokens(&self) {
+        println!("tokens: {:?}", self.tokens);
```

Comment:
For some reason, I'm currently not able to iterate over `Vec<Token>`. Maybe I need to experiment with turning it into an `iter` before using it in algs.